src/main/java/carpet/script/annotation/AnnotationParser.java

```diff
@@ -0,0 +1,318 @@
+                            + "Provided in " + instance.getClass());
+                if (maxParams == -1)
+                    maxParams = Integer.MAX_VALUE;
+            }
+            this.maxParams = maxParams;
+
+            // Why MethodHandles?
+            // MethodHandles are blazing fast (in some situations they can even compile to a single invokeVirtual), but slightly complex to work with.
+            // Their "polymorphic signature" makes them (by default) require the exact signature of the method and return type, in order to call the
+            // functions directly, not even accepting Objects. Therefore we change them to spreaders (accept array instead), return type of Object,
+            // and we also bind them to our instance. That makes them ever-so-slightly slower, since they have to cast the params, but it's not
+            // noticeable, and comparing the overhead that reflection's #invoke had over this, the difference is substantial
+            // (checking access at invoke vs create).
+            // Note: there is also MethodHandle#invoke and #invokeWithArguments, but those run #asType at every invocation, which is quite slow, so we
+            // are basically running it here.
+            try
+            {
+                MethodHandle tempHandle = MethodHandles.publicLookup().unreflect(method).asFixedArity().asSpreader(Object[].class, this.methodParamCount);
+                this.handle = tempHandle.asType(tempHandle.type().changeReturnType(Object.class)).bindTo(instance);
+            } catch (IllegalAccessException e)
+            {
+                throw new IllegalArgumentException(e);
+            }
+
+            this.scarpetParamCount = this.isEffectivelyVarArgs ? -1 : this.minParams;
+            this.contextType = method.getAnnotation(ScarpetFunction.class).contextType();
+        }
+
+        @Override
+        public LazyValue apply(Context context, Context.Type t, List<LazyValue> lazyValues)
+        {
+            List<Value> lv = AbstractLazyFunction.unpackLazy(lazyValues, context, contextType);
+            if (isEffectivelyVarArgs)
+            {
+                if (lv.size() < minParams)
+                    throw new InternalExpressionException(name + " expected at least " + minParams + " arguments, got " + lv.size() + ". " + getUsage());
+                if (lv.size() > maxParams)
+                    throw new InternalExpressionException( name + " expected up to " + maxParams + " arguments, got " + lv.size() + ". " + getUsage());
+            } else
+            {
+                if (lv.size() != minParams) // min == max if not varargs. We do this cause we don't use function's arg checking to be able to return lazy
+                    throw new InternalExpressionException("Function " + name + " expected " + minParams + " arguments, got " + lv.size() + ". " + getUsage());
```

Comment:
'most' error messages (I like to be 'all') when indicating a faulty method, use 'quotes' to highlight the faulty API call.
Consider `Function incorrectly expected ....` vs. `Function 'incorrectly' expected ... `
```suggestion
                    throw new InternalExpressionException("Function '" + name + "' expected " + minParams + " arguments, got " + lv.size() + ". " + getUsage());
```

Reply:
Whoops, didn't notice this was a click to fix, implemented it from the IDE (same for the other two, also found a few more).